src/game/client/baseheightmap.cpp

```diff
@@ -13,12 +13,459 @@
+    Ref_Ptr_Release(m_stageOneTexture);
+    Ref_Ptr_Release(m_stageTwoTexture);
+    Ref_Ptr_Release(m_stageThreeTexture);
+    Ref_Ptr_Release(m_destAlphaLUT);
+    Ref_Ptr_Release(m_map);
+    return 0;
+}
+
+void BaseHeightMapRenderObjClass::Release_Resources()
+{
+    if (m_treeBuffer != nullptr) {
+        m_treeBuffer->Free_Tree_Buffers();
+    }
+
+    if (m_bibBuffer != nullptr) {
+        m_bibBuffer->Free_Bib_Buffers();
+    }
+
+    if (m_bridgeBuffer != nullptr) {
+        m_bridgeBuffer->Free_Bridge_Buffers();
+    }
+
+    if (m_waypointBuffer != nullptr) {
+        m_waypointBuffer->Free_Waypoint_Buffers();
+    }
+
+    if (m_map != nullptr) {
+        m_map->Add_Ref();
+    }
+
+    WorldHeightMap *map = m_map;
+    Free_Map_Resources();
+    m_map = map;
+
+    if (g_theWaterRenderObj != nullptr) {
+        g_theWaterRenderObj->Release_Resources();
+    }
+
+    if (g_theTerrainTracksRenderObjClassSystem != nullptr) {
+        g_theTerrainTracksRenderObjClassSystem->Release_Resources();
+    }
+
+    if (g_theW3DShadowManager != nullptr) {
+        g_theW3DShadowManager->Release_Resources();
+    }
+
+    if (m_shroud != nullptr) {
+        m_shroud->Reset();
+        m_shroud->Release_Resources();
+    }
+
+    if (g_theSmudgeManager != nullptr) {
+        g_theSmudgeManager->Release_Resources();
+    }
+
+    if (g_theSnowManager != nullptr) {
+        g_theSnowManager->Release_Resources();
+    }
+
+    W3DShaderManager::Shutdown();
+
+    if (m_roadBuffer != nullptr) {
+        m_roadBuffer->Free_Road_Buffers();
+    }
+}
+
+void BaseHeightMapRenderObjClass::Re_Acquire_Resources()
+{
+    W3DShaderManager::Init();
+
+    if (g_theWaterRenderObj != nullptr) {
+        g_theWaterRenderObj->Re_Acquire_Resources();
+    }
+
+    if (g_theTerrainTracksRenderObjClassSystem != nullptr) {
+        g_theTerrainTracksRenderObjClassSystem->Re_Acquire_Resources();
+    }
+
+    if (g_theW3DShadowManager != nullptr) {
+        g_theW3DShadowManager->Re_Acquire_Resources();
+    }
+
+    if (m_shroud != nullptr) {
+        m_shroud->Re_Acquire_Resources();
+    }
+
+    if (m_map != nullptr) {
+        Init_Height_Data(m_x, m_y, m_map, nullptr, true);
+        m_needFullUpdate = 1;
+    }
+
+    if (m_treeBuffer != nullptr) {
+        m_treeBuffer->Allocate_Tree_Buffers();
+    }
+
+    if (m_bibBuffer != nullptr) {
+        m_bibBuffer->Allocate_Bib_Buffers();
+    }
+
+    if (m_bridgeBuffer != nullptr) {
+        m_bridgeBuffer->Allocate_Bridge_Buffers();
+    }
+
+    if (g_theSmudgeManager != nullptr) {
+        g_theSmudgeManager->Re_Acquire_Resources();
+    }
+
+    if (g_theSnowManager != nullptr) {
+        g_theSnowManager->Re_Acquire_Resources();
+    }
+
+    if (m_roadBuffer != nullptr) {
+        m_roadBuffer->Allocate_Road_Buffers();
+        m_roadBuffer->Load_Roads();
+    }
+    if (g_theTacticalView != nullptr) {
+        g_theTacticalView->Force_Redraw();
+        g_theTacticalView->Force_Redraw(); // possible bug

```

Comment:
Add a "TODO" keyword to make it discoverable.

Reply:
fixed.